src/spawn_canister/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use ic_cdk::export::candid::{CandidType, Deserialize, Principal, Encode};
+use ic_cdk_macros::{update};
+use ic_cdk::api::call;
+
+#[derive(CandidType, Deserialize)]
+pub enum CreateCanisterResponse{
+    #[serde(rename = "created")]
+    Created(Principal),
+
+    #[serde(rename = "insufficient_funds")]
+    InsufficientFunds,
+
+    #[serde(rename = "canister_creation_error")]
+    CanisterCreationError,
+
+    #[serde(rename = "canister_installation_error")]
+    CanisterInstallationError,
+
+    #[serde(rename = "change_controller_error")]
+    ChangeControllerError,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct CreateCanisterResult {
+    pub canister_id: Principal,
+}
+
+#[derive(CandidType, Deserialize)]
+pub enum InstallMode {
+    #[serde(rename = "install")]
+    Install,
+    #[serde(rename = "reinstall")]
+    Reinstall,
+    #[serde(rename = "upgrade")]
+    Upgrade,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct InstallCodeArg {
+    pub mode: InstallMode,
+    pub canister_id: Principal,
+    pub wasm_module: Vec<u8>,
+    pub arg : Vec<u8>,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct CanisterSettings{
+    controllers: Option<Vec<Principal>>,
+    compute_allocation: Option<candid::Nat>,
+    memory_allocation: Option<candid::Nat>,
+    freezing_threshold: Option<candid::Nat>,
+}
+
+#[derive(CandidType, Deserialize)]
+pub struct UpdateSettingsArg{
+    canister_id: Principal,
+    settings: CanisterSettings,
+}
+
+static VIDEO_CANISTER_CODE: &[u8;  include_bytes!("../../../target/wasm32-unknown-unknown/release/video_canister_opt.wasm").len()] = include_bytes!("../../../target/wasm32-unknown-unknown/release/video_canister_opt.wasm");
+
+const MIN_CANISTER_CYCLES_REQUIRED: u64 = 100_000_000_000 + 10_000_000_000; //TODO rough guess, calculate correct costs
+
+#[update]
+pub async fn create_new_canister() -> CreateCanisterResponse{
+
+    let owner = ic_cdk::api::caller(); //call early before any callbacks from calling other canisters
+    let available_cycles = call::msg_cycles_available();  
+
+    if available_cycles < MIN_CANISTER_CYCLES_REQUIRED{
+        return CreateCanisterResponse::InsufficientFunds;
+    } else {
+        call::msg_cycles_accept(available_cycles);
+    }
+
+    let canister_princ = match create_canister_on_network().await{
+        Ok(new_princ) => new_princ,
+        Err(_err_str) => {
+            return CreateCanisterResponse::CanisterCreationError;
+        }
+    };
+
+    if let Err(_err_str) = install_video_canister(canister_princ.clone(), &owner).await{
+        return CreateCanisterResponse::CanisterInstallationError;
+    };
+
+    if let Err(_err_str) = change_controller_to_owner(canister_princ.clone(), &owner).await{
+        return CreateCanisterResponse::ChangeControllerError;
+    }
+
+    return CreateCanisterResponse::Created(canister_princ);
+}
+
+async fn create_canister_on_network() -> Result<Principal, String>{
```

Comment:
gefuehlt in den 2 create methoden 3 verschiedene stile um mit asynchronen calls umzugehen. Vereinheitlichung moeglich?

Reply:
Bitte Klarifizieren, da weis ich nicht genau was du meinst

Reply:
Vergleiche Zeile 76, 83, 87, 97.
Einmal matched du die response im await, einmal die rückgabe direkt, einmal speicherst du die rückgabe in einer variable und matcht dann. ist nicht weiter tragisch, schaut nur komisch aus. eventuell ist es auch der schönstmögliche weg den du gemacht hast für den zweck

Reply:
76 ist anders wegen anderer return struktur,
83,87 könnte man auch so wie 76 schreiben da wäre der Code aber dann genauso lang mit logisch leeren Zeilen, finde ich persönlich unschön
97 braucht der Compiler den Typen damit das call result dann auf den richtigen Typen geparsed wird, man könnte das auch bei den anderen Zeilen schreiben, finde ich persönlich aber wieder unschön wenn nicht nötig